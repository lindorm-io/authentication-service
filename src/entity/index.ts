export * from "./Account";
export * from "./Client";
export * from "./Device";
export * from "./RequestLimit";
export * from "./Session";
