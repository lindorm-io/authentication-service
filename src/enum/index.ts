export * from "./account-event";
export * from "./audience";
export * from "./device-event";
export * from "./grant-type";
export * from "./mongo-collection";
export * from "./multi-factor-challenge-type";
export * from "./response-type";
export * from "./session-event";
