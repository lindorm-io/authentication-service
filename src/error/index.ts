export * from "./AccountNotFoundError";
export * from "./AssertClientSecretError";
export * from "./AssertCodeChallengeError";
export * from "./AssertDeviceChallengeError";
export * from "./AssertOTPError";
export * from "./DeviceNotFoundError";
export * from "./ExpiredSessionError";
export * from "./InvalidAuthorizationError";
export * from "./InvalidAuthorizationTokenError";
export * from "./InvalidClientError";
export * from "./InvalidDeviceError";
export * from "./InvalidExpiryString";
export * from "./InvalidGrantTypeError";
export * from "./InvalidPermissionError";
export * from "./InvalidRefreshTokenError";
export * from "./InvalidScopeError";
export * from "./InvalidSubjectError";
export * from "./LockedAccountError";
export * from "./RejectedClientError";
export * from "./RequestLimitBackOffError";
