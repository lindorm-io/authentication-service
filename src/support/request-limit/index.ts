export * from "./create-or-update-request-limit";
export * from "./validate-back-off";
