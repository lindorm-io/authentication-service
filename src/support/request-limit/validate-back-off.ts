import { IAuthContext } from "../../typing";
import { isBefore } from "date-fns";

export const validateRequestLimitBackOff = (ctx: IAuthContext) => (): void => {
  const { requestLimit } = ctx;

  if (!requestLimit?.backOffUntil) return;

  if (isBefore(new Date(), ctx.requestLimit.backOffUntil)) {
    throw new Error("Rate Limited");
  }
};
