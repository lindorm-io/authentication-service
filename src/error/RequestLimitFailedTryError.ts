import { APIError, ExtendableError, HttpStatus } from "@lindorm-io/core";
import { RequestLimit } from "../entity";

export class RequestLimitFailedTryError extends APIError {
  constructor(requestLimit: RequestLimit, error: ExtendableError) {
    super("This request failed and has been rate-limited", {
      publicData: {
        failedTries: requestLimit.failedTries,
        backOffUntil: requestLimit.backOffUntil,
      },
      details: error?.details || error?.message,
      debug: { error },
      statusCode: HttpStatus.ClientError.FORBIDDEN,
    });
  }
}
