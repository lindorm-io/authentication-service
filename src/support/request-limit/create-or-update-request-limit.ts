import { GrantType } from "../../enum";
import { IAuthContext } from "../../typing";
import { RequestLimit } from "../../entity";
import { getBackOffDate } from "./back-off";

export interface ICreateOrUpdateOptions {
  grantType: GrantType;
  subject: string;
}

export const createOrUpdateRequestLimit = (ctx: IAuthContext) => async (
  options: ICreateOrUpdateOptions,
): Promise<void> => {
  const { cache, requestLimit } = ctx;
  const { grantType, subject } = options;

  if (!requestLimit) {
    await cache.requestLimit.create(
      new RequestLimit({
        subject,
        grantType,
      }),
    );

    return;
  }

  ctx.requestLimit.failedTries = requestLimit.failedTries + 1;
  ctx.requestLimit.backOffUntil = getBackOffDate(requestLimit);
};
