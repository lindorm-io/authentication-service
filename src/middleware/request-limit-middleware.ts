import Joi from "@hapi/joi";
import { IAuthContext } from "../typing";
import { JOI_EMAIL, JOI_GRANT_TYPE } from "../constant";
import { RequestLimitCache } from "../cache";
import { TPromise } from "@lindorm-io/core";
import { createOrUpdateRequestLimit, validateRequestLimitBackOff } from "../support";
import { CacheEntityNotFoundError } from "@lindorm-io/redis/dist/error";

const schema = Joi.object({
  grantType: JOI_GRANT_TYPE,
  subject: JOI_EMAIL,
});

export const requestLimitMiddleware = async (ctx: IAuthContext, next: TPromise<void>): Promise<void> => {
  const start = Date.now();

  const { cache, logger } = ctx;
  const { grantType, subject } = ctx.request.body;

  await schema.validateAsync({ grantType, subject });

  try {
    ctx.requestLimit = await cache.requestLimit.find(RequestLimitCache.getKey({ grantType, subject }));

    logger.info("request is rate-limited");
  } catch (err) {
    if (err instanceof CacheEntityNotFoundError) {
      logger.info("request is not rate-limited");
    } else {
      throw err;
    }
  }

  ctx.metrics = {
    ...(ctx.metrics || {}),
    requestLimit: Date.now() - start,
  };

  if (ctx.requestLimit) {
    validateRequestLimitBackOff(ctx)();
  }

  try {
    await next();
  } catch (err) {
    await createOrUpdateRequestLimit(ctx)({ grantType, subject });

    throw err;
  }
};
