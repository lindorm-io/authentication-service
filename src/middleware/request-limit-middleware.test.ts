import * as create from "../support/request-limit/create-or-update-request-limit";
import * as validate from "../support/request-limit/validate-back-off";
import MockDate from "mockdate";
import { GrantType } from "../enum";
import { MOCK_LOGGER, mockCacheRequestLimit } from "../test/mocks";
import { requestLimitMiddleware } from "./request-limit-middleware";
import { CacheEntityNotFoundError } from "@lindorm-io/redis/dist/error";
import { RequestLimit } from "../entity";
import { AssertDeviceChallengeError } from "../error";
import { RequestLimitFailedTryError } from "../error/RequestLimitFailedTryError";

jest.mock("uuid", () => ({
  v4: jest.fn(() => "be3a62d1-24a0-401c-96dd-3aff95356811"),
}));

MockDate.set("2020-01-01 08:00:00.000");

describe("requestLimitMiddleware", () => {
  let ctx: any;
  let next: any;

  let spyCreateOrUpdateRequestLimit: any;
  let spyValidateRequestLimitBackOff: any;

  beforeEach(() => {
    ctx = {
      cache: {
        requestLimit: mockCacheRequestLimit,
      },
      logger: MOCK_LOGGER,
      request: {
        body: {
          grantType: GrantType.DEVICE_PIN,
          subject: "subject@lindorm.io",
        },
      },
    };
    next = () => Promise.resolve();

    spyCreateOrUpdateRequestLimit = jest
      // @ts-ignore
      .spyOn(create, "createOrUpdateRequestLimit")
      .mockImplementation(() => () => undefined);

    spyValidateRequestLimitBackOff = jest
      // @ts-ignore
      .spyOn(validate, "validateRequestLimitBackOff")
      .mockImplementation(() => () => undefined);
  });

  afterEach(jest.resetAllMocks);

  test("should consider request rate limited", async () => {
    await expect(requestLimitMiddleware(ctx, next)).resolves.toBe(undefined);

    expect(ctx.requestLimit).toMatchSnapshot();
    expect(ctx.metrics.requestLimit).toStrictEqual(expect.any(Number));

    expect(spyCreateOrUpdateRequestLimit).not.toHaveBeenCalled();
    expect(spyValidateRequestLimitBackOff).toHaveBeenCalled();
  });

  test("should consider request clean", async () => {
    ctx.cache.requestLimit.find = jest.fn(() => {
      throw new CacheEntityNotFoundError("key", { result: true });
    });

    await expect(requestLimitMiddleware(ctx, next)).resolves.toBe(undefined);

    expect(ctx.requestLimit).toMatchSnapshot();
    expect(ctx.metrics.requestLimit).toStrictEqual(expect.any(Number));

    expect(spyValidateRequestLimitBackOff).not.toHaveBeenCalled();
  });

  test("should create or update request limit", async () => {
    ctx.requestLimit = new RequestLimit({
      grantType: GrantType.DEVICE_PIN,
      subject: "test@lindorm.io",
    });
    next = () => Promise.reject(new AssertDeviceChallengeError("string", "verifier"));

    await expect(requestLimitMiddleware(ctx, next)).rejects.toThrow(expect.any(RequestLimitFailedTryError));

    expect(spyCreateOrUpdateRequestLimit).toHaveBeenCalled();
  });

  test("should throw error", async () => {
    ctx.cache.requestLimit.find = jest.fn(() => {
      throw new Error("mock");
    });

    await expect(requestLimitMiddleware(ctx, next)).rejects.toStrictEqual(new Error("mock"));

    expect(spyValidateRequestLimitBackOff).not.toHaveBeenCalled();
  });
});
