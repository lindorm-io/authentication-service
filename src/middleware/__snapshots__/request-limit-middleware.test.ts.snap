// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`requestLimitMiddleware should consider request clean 1`] = `undefined`;

exports[`requestLimitMiddleware should consider request rate limited 1`] = `
RequestLimit {
  "_backOffUntil": null,
  "_created": 2020-01-01T07:00:00.000Z,
  "_events": Array [],
  "_failedTries": 1,
  "_grantType": "pin_code",
  "_id": "be3a62d1-24a0-401c-96dd-3aff95356811",
  "_subject": "pin_code::subject@lindorm.io",
  "_updated": 2020-01-01T07:00:00.000Z,
  "_version": 0,
}
`;
